lib/portlayer/storage/vsphere/datastore.go

```diff
@@ -0,0 +1,243 @@
+// Copyright 2016 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package vsphere
+
+import (
+	"io"
+	"path"
+
+	log "github.com/Sirupsen/logrus"
+	"github.com/vmware/govmomi/object"
+	"github.com/vmware/govmomi/vim25/mo"
+	"github.com/vmware/govmomi/vim25/soap"
+	"github.com/vmware/govmomi/vim25/types"
+	"github.com/vmware/vic/pkg/vsphere/session"
+	"github.com/vmware/vic/pkg/vsphere/tasks"
+	"golang.org/x/net/context"
+)
+
+// Datastore gives access to the datastore regardless of type (esx, esx + vc,
+// or esx + vc + vsan).  Also wraps paths to a given root directory
+type datastore struct {
+	// The Datastore API likes everything in "path/to/thing" format.
+	ds *object.Datastore
+
+	s *session.Session
+
+	// The FileManager API likes everything in "[dsname] path/to/thing" format.
+	fm *object.FileManager
+
+	// the name of the datastore path (in case of vsan, this is the human readable name)
+	rootname string
+
+	// The datastore url (including root)
+	rooturl string
+
+	// datastore path (not in url form)
+	rootdir string
+}
+
+// newDatastore returns a Datastore.
+// ctx is a context,
+// s is an authenticated session
+// ds is the vsphere datastore
+// root is the top level directory to root all data.  If root does not exist,
+// it will be created.  If it already exists, NOOP. This cannot be empty.
+func newDatastore(ctx context.Context, s *session.Session, ds *object.Datastore, root string) (*datastore, error) {
+	d := &datastore{
+		ds:       ds,
+		s:        s,
+		rootname: root,
+		fm:       object.NewFileManager(s.Vim25()),
+	}
+
+	if err := d.mkRootDir(ctx, root); err != nil {
+		log.Debugf("error creating root directory %s: %s", root, err)
+		return nil, err
+	}
+
+	log.Infof("Datastore path is %s", d.rooturl)
+	return d, nil
+}
+
+func (d *datastore) Summary(ctx context.Context) (*types.DatastoreSummary, error) {
+
+	var mds mo.Datastore
+	if err := d.ds.Properties(ctx, d.ds.Reference(), []string{"info", "summary"}, &mds); err != nil {
+		return nil, err
+	}
+
+	return &mds.Summary, nil
+}
+
+// Mkdir creates directories.  The force flag creates the dir even if it exists. Returns the datastore url of the created directory
+func (d *datastore) Mkdir(ctx context.Context, force bool, dirs ...string) (string, error) {
```

Comment:
Good catch!  It should be `createParentDirectories` (not `force`) and it should be passed through to the `fm.MakeDirectory` call.  Will fix.